modules/login/php/passwordchange.class.inc

```diff
@@ -0,0 +1,170 @@
+<?php
+/**
+ * Implements the user password reset page for LORIS.
+ *
+ * PHP version 5
+ *
+ * @category   Behavioural
+ * @package    Main
+ * @subpackage Imaging
+ * @author     Dave MacFarlane <david.macfarlane2@mcgill.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link       https://www.github.com/aces/Loris/
+ */
+namespace LORIS\login;
+
+/**
+ * Implements the user password reset page for LORIS.
+ *
+ * @category   Behavioural
+ * @package    Main
+ * @subpackage Imaging
+ * @author     Dave MacFarlane <david.macfarlane2@mcgill.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link       https://www.github.com/aces/Loris/
+ */
+class PasswordChange extends \NDB_Form
+{
+    /**
+     * Determine whether the user has permission to view this page
+     *
+     * @return bool whether the user has access
+     */
+    function _hasAccess()
+    {
+        return true;
+    }
+
+    /**
+     * CommonPageSetup sets up the common form elements which
+     * are on the page, regardless of HTTP request method.
+     *
+     * @return none
+     */
+    function setup()
+    {
+        $this->tpl_data['page_title'] = 'Update Password';
+        $factory  = \NDB_Factory::singleton();
+        $settings = $factory->settings();
+        $baseURL  = $settings->getBaseURL();
+        $DB       =& \Database::singleton();
+
+        $token = $_GET['token'] ?? '';
+
+        $validToken = $DB->pselectRow(
+            'SELECT *, TIMESTAMPDIFF(HOUR,time,CURRENT_TIMESTAMP) as token_age 
+              FROM password_recovery 
+              WHERE token=:tok',
+            array('tok' => $token)
+        );
+
+        if (empty($token) || empty($validToken)) {
+            header("Location: {$baseURL}/");
+        }
+
+        $username = \User::getUsernameFromId($validToken['user_id']);
+
+        $this->addPassword(
+            "password",
+            "",
+            array(
+             "placeholder" => "New Password",
+             "class"       => "form-control input-sm",
+            )
+        );
+        $this->addPassword(
+            "confirm",
+            "",
+            array(
+             "placeholder" => "Confirm Password",
+             "class"       => "form-control input-sm",
+            )
+        );
+
+        $this->addHidden('token', $_GET['token']);
+        $this->addHidden('username', $username);
+        $this->form->addFormRule(array(&$this, '_validate'));
```

Comment:
@driusan shouldn't this be done automatically on all forms ??

without this line, the _validate function is never called

Reply:
I think that would make sense, but we have a lot of pages which aren't forms that are extending NDB_Form instead of NDB_Page and I think it should be investigated separately from this PR...